core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java

```diff
@@ -145,11 +146,29 @@ public BaseUpdatePartitionSpec addField(String name, Term term) {
     checkForRedundantAddedPartitions(newField);
 
     transformToAddedField.put(validationKey, newField);
-    if (name != null) {
-      nameToAddedField.put(name, newField);
+
+    String partitionName;
+    if (newField.name() != null) {
+      partitionName = newField.name();
+    } else {
+      partitionName = PartitionSpecVisitor.visit(schema, newField, PartitionNameGenerator.INSTANCE);
     }
 
-    adds.add(newField);
+    PartitionField existingField = nameToField.get(partitionName);
+    if (existingField != null) {
+      if (isVoidTransform(existingField)) {
+        // rename the old deleted field that is being replaced by the new field
+        renameField(existingField.name(), existingField.name() + "_" + UUID.randomUUID());
```

Comment:
Here as well, I'd prefer not to use a UUID. This should be able to use the existing field's ID instead.

Reply:
use the existing field's ID instead.